processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,481 @@
+    // * https://stackoverflow.com/questions/11761552/detailed-semantics-of-volatile-regarding-timeliness-of-visibility
+
+    if (curWriteIndex == maxNumSlots - 1) {
+      // We additionally check that nextReadIndex is -1 here because the writing thread should wait for the reading
+      // thread to start reading only when the writing thread tries to overwrite the first slot for the first time.
+      while ((nextReadIndex == -1 || nextReadIndex == 0) && !Thread.currentThread().isInterrupted()) {
+        if (now >= queryTimeoutAtNs) {
+          throw new RuntimeException(new TimeoutException());
+        }
+        // Thread.yield() should not be called from the very beginning
+        if (now >= spinTimeoutAt) {
+          Thread.yield();
+        }
+        now = System.nanoTime();
+      }
+      curWriteIndex = 0;
+    } else {
+      final int nextWriteIndex = curWriteIndex + 1;
+      while ((nextWriteIndex == nextReadIndex) && !Thread.currentThread().isInterrupted()) {
+        if (now >= queryTimeoutAtNs) {
+          throw new RuntimeException(new TimeoutException());
+        }
+        // Thread.yield() should not be called from the very beginning
+        if (now >= spinTimeoutAt) {
+          Thread.yield();
+        }
+        now = System.nanoTime();
+      }
+      curWriteIndex = nextWriteIndex;
+    }
+  }
+
+  @Override
+  public void reset()
+  {
+    curWriteIndex = -1;
+    nextReadIndex = -1;
+    finished = false;
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      try {
+        aggregator.close();
+      }
+      catch (Exception e) {
+        LOG.warn(e, "Could not close aggregator [%s], skipping.", aggregator);
+      }
+    }
+  }
+
+  /**
+   * Signal that no more inputs are added.  Must be called after {@link #aggregate(Object)} is called for the last input.
+   */
+  public void finish()
+  {
+    increaseWriteIndex();
+    finished = true;
+  }
+
+  /**
+   * Return a sorted iterator.  This method can be called safely while writing, and the iterating thread and the writing
+   * thread can be different.  The result iterator always returns sorted results.  This method should be called only one
+   * time per grouper.
+   *
+   * @return a sorted iterator
+   */
+  public CloseableIterator<Entry<KeyType>> iterator()
+  {
+    if (!initialized) {
+      throw new ISE("Grouper should be initialized first");
+    }
+
+    return new CloseableIterator<Entry<KeyType>>()
+    {
+      {
+        // Wait for some data to be ready and initialize nextReadIndex.
+        increaseReadIndexTo(0);
+      }
+
+      @Override
+      public boolean hasNext()
+      {
+        // Once finished becomes true, curWriteIndex isn't changed anymore and thus remainig() can be computed safely
+        // because nextReadIndex is changed only by the reading thread.
+        return !finished || remaining() > 0;
```

Comment:
Here should be a mention of happens-before edge on `finished`

Reply:
Added.